File: electron/preload-tauri/ready.ts
// JS side of the "frontend-ready" event (Hearth#27 Phase 1's IPC scope),
// backed by src-tauri/src/ready.rs's `frontend_ready` command.
//
// Call this once the renderer has actually mounted (not just once the HTML
// document parsed) — see ready.rs's doc comment for why that distinction
// matters to the boot watchdog. There's no Electron-side equivalent to
// mirror here: the original relies on `did-finish-load`, a native
// webContents event with no renderer-side call at all.

interface TauriGlobal {
  core: {
    invoke: (cmd: string, args?: Record<string, unknown>) => Promise<unknown>
  }
}

export function frontendReady(): Promise<void> {
  const g = (window as unknown as { __TAURI__?: TauriGlobal }).__TAURI__
  if (!g) {
    throw new Error('preload-tauri/ready: window.__TAURI__ is not present (is withGlobalTauri enabled?)')
  }
  return g.core.invoke('frontend_ready') as Promise<void>
}

File: src-tauri/src/ready.rs
// The "new frontend-ready event" called for by Hearth#27 Phase 1's IPC scope.
//
// TS confirms the boot watchdog once the window reaches Electron's
// `did-finish-load` — a native webContents event that fires once the HTML
// document has loaded. That's a coarser signal than we want: it fires once
// the page is *parsed*, not once the renderer has actually mounted and run
// without throwing. A renderer that crashes during React mount would still
// have fired `did-finish-load` moments earlier, so the watchdog would wrongly
// consider the boot healthy.
//
// This command inverts that: the renderer itself calls `frontendReady()`
// once mounted, so a boot that crashes before mounting never confirms —
// exactly the case the watchdog exists to catch.

use crate::selfmod_commands::AppState;

#[tauri::command]
pub fn frontend_ready(state: tauri::State<AppState>) {
    state.boot_watchdog.confirm_ready();
}
